# The scrapping part

## *First of all we are going to create a litlle scrapping programms.*

In the first places import webdriver from selenium.

And connect a driver variable to this webpage : "https://finance.yahoo.com/most-active?offset=0&count=100"

In [1]:
from selenium import webdriver


driver = webdriver.Firefox()
driver.get("https://finance.yahoo.com/most-active?offset=0&count=100")

Firefox should open a webpage, don't close it. Once the webpage is open, you should be able to see a button to accept cookies.

To complete this task, you will need to import By from the Selenium library and use it to find the cookies button element in the webpage.

Tips : Don't forget to click the button, and you'll maybe need to find the scroll down button !

In [2]:
from selenium.webdriver.common.by import By

scroll_button = driver.find_element(By.ID, "scroll-down-btn")
scroll_button.click()

cookie_button = driver.find_element(By.XPATH, "//button[@class='btn secondary accept-all ' and @name='agree']")
cookie_button.click()

Now that the cookies have been accepted, it's time for scraping!

First of all, to create our CSV file, we will need the header. On the page, you can see that there is a header at the top of the stock data. Please retrieve the header from "Name" to "PE Ratio (TTM)" included. 

Create a list with the header and name it : my_header_list

In [3]:
thead_element = driver.find_element(By.TAG_NAME,"thead")
th_elem = thead_element.find_elements(By.TAG_NAME,"th")
my_header_list = []
for elem in th_elem[:-1] : 
    my_header_list.append(elem.text)

Print my_header_list

In [4]:
print(my_header_list)

['Symbol', 'Name', 'Price (Intraday)', 'Change', '% Change', 'Volume', 'Avg Vol (3 month)', 'Market Cap', 'PE Ratio (TTM)']


Expected : ['Symbol', 'Name', 'Price (Intraday)', 'Change', '% Change', 'Volume', 'Avg Vol (3 month)', 'Market Cap', 'PE Ratio (TTM)']

It is now time to get the stock data. Since we have the header, we need to correlate it with the actual data. Please retrieve the 100 stock data points that are on the page and store them in a list called "entreprise_stock_info_list".

Tip: "entreprise_stock_info_list" is a list that contains multiple lists! 

In [5]:
tbody_element = driver.find_element(By.TAG_NAME,"tbody")
tr_elem_from_tbody  = tbody_element.find_elements(By.TAG_NAME,"tr")
entreprise_stock_info_list = []
for selenium_elem_tr in tr_elem_from_tbody :
    td_elem_in_tr = selenium_elem_tr.find_elements(By.TAG_NAME,"td")
    entreprise_stock_info_list.append([])
    for selenium_elem_td in td_elem_in_tr :
        entreprise_stock_info_list[-1].append(selenium_elem_td.text)

Now print : entreprise_stock_info_list

In [6]:
print(entreprise_stock_info_list)

[['CS', 'Credit Suisse Group AG', '2.1600', '-0.3500', '-13.94%', '422.265M', '29.199M', '9.671B', 'N/A', ''], ['TSLA', 'Tesla, Inc.', '180.45', '-2.81', '-1.53%', '145.996M', '184.604M', '579.564B', '47.86', ''], ['BAC', 'Bank of America Corporation', '28.49', '-0.27', '-0.94%', '130.95M', '45.96M', '227.899B', '8.85', ''], ['F', 'Ford Motor Company', '11.71', '-0.22', '-1.84%', '107.025M', '65.232M', '47.555B', 'N/A', ''], ['FRC', 'First Republic Bank', '31.16', '-8.47', '-21.37%', '95.937M', '8.409M', '5.803B', '2.98', ''], ['SCHW', 'The Charles Schwab Corporation', '59.55', '+2.87', '+5.06%', '86.538M', '15.231M', '108.912B', '15.59', ''], ['AMD', 'Advanced Micro Devices, Inc.', '89.68', '+2.23', '+2.55%', '86.177M', '57.046M', '144.509B', '103.08', ''], ['HBAN', 'Huntington Bancshares Incorporated', '10.85', '-0.20', '-1.81%', '71.166M', '16.441M', '15.729B', '7.59', ''], ['ITUB', 'Itaú Unibanco Holding S.A.', '4.4100', '-0.0700', '-1.56%', '65.813M', '31.57M', '43.107B', '7.60', 

Expected : [['CS', 'Credit Suisse Group AG', '2.1600', '-0.3500', '-13.94%', '422.265M', '29.199M', '9.671B', 'N/A', ''], ['TSLA', 'Tesla, Inc.', '180.45', '-2.81', '-1.53%', '145.996M', '184.604M', '579.564B', '47.86', ''], . . .]

You can see that we have an unexpected element at the end of each list. We'll remove it later don't worry !

Let's now close and quit our driver

In [7]:
driver.close()
driver.quit()

__Scrapping finished__

*Time to create the CSV*

First step : Import pandas as pd
Create a pandas dataframe with the entreprise_stock_info_list and my_header_list

Tips : Don't forget to remove the last elements of each inner list in entreprise_stock_info_list

In [8]:
import pandas as pd
for elem in entreprise_stock_info_list :
        elem.pop()  # remove the last element of each inner list
df = pd.DataFrame(data=entreprise_stock_info_list, columns=my_header_list)

Let's now convert it to a CSV -> data/simple_data.csv

tips : ../data/simple_data.csv

In [11]:
df.to_csv("../data/simple_data.csv", index=False)

<img src="..\img\simpla_data_img.png" alt="Alternative text" />